lib/streamlit/connections/base_connection.py

```diff
@@ -12,15 +12,67 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from abc import ABC
-from typing import Generic, TypeVar
+from abc import ABC, abstractmethod
+from typing import Any, Generic, Mapping, Optional, TypeVar
+
+from streamlit.runtime.secrets import AttrDict, secrets_singleton
 
 T = TypeVar("T")
 
 
 class BaseConnection(ABC, Generic[T]):
-    """TODO(vdonato): Implement... this entire class.
+    """TODO(vdonato): docstrings for this class and all public methods."""
+
+    def __init__(self, connection_name: str = "default", **kwargs) -> None:
+        self._connection_name = connection_name
+        self._kwargs = kwargs
+
+        self._raw_instance: Optional[T] = self.connect(**kwargs)
+
+    def _repr_html_(self) -> str:
+        # TODO(vdonato): Change this to whatever we actually want the default to be.
+        return f"Hi, I am a {self.default_connection_name()} connection!"
+
+    # Methods with default implementations that we don't expect subclasses to want or
+    # need to overwrite.
+    def get_secrets(self) -> Mapping[str, Any]:
+        connection_name = self._connection_name
+        if connection_name == "default":
+            connection_name = self.default_connection_name()
+
+        connections_section = None
+        if secrets_singleton.load_if_toml_exists():
+            connections_section = secrets_singleton.get("connections")
+
+        if type(connections_section) is not AttrDict:
+            return AttrDict({})
+
+        return connections_section.get(connection_name, {})
+
+    @classmethod
+    def default_connection_name(cls) -> str:
+        name = cls._default_connection_name
+
+        if name is None:
+            raise NotImplementedError(
+                "Subclasses of BaseConnection must define a _default_connection_name attribute."
+            )
+        return name
+
+    # TODO(vdonato): Finalize the name for this method. Should this be `invalidate`?
+    def reset(self) -> None:
```

Comment:
nit: same as above. If this is part of the public API, we could add `@gather_metrics("connection:reset")`

Reply:
This one is another one that I don't think it'd be too useful to collect metrics on.

It's part of the public API and is meant to be used by people writing apps, but it's intended usage is to be able to help implement error handling / retries, so I think we'd just end up indirectly measuring connection error rates in various Streamlit app deployments if we're collecting metrics for this. Will also check with the data team on this, though